traits/ctraits.c

```diff
@@ -310,10 +310,39 @@ fatal_trait_error ( void ) {
 +----------------------------------------------------------------------------*/
 
 static int
-invalid_attribute_error ( void ) {
+invalid_attribute_error ( PyObject * name ) {
 
-    PyErr_SetString( PyExc_TypeError, "attribute name must be string" );
+    // Build error message by collecting the type of the object received
+    PyTypeObject* obj_type = name->ob_type;
+    const char* type_name = obj_type->tp_name;
+    PyObject * ob_repr = PyObject_Repr( name );
 
+#if PY_MAJOR_VERSION >= 3
+
+    const char* fmt = "attribute name must be an instance of <type 'str'>. "
+                      "Got %R (%.200s).";
```

Comment:
This still looks wrong: we're computing the `repr` of `name`, and then showing the `repr` of that `repr`! I'll rework.